# 05-异步编程 -3-生成器

## 一 理解生成器

生成器是 ES6 新增的结构，拥有的能力：**在一个函数块内暂停、恢复代码的执行**。

生成器的原理是其内部被包装成了迭代器，利用 next() 的方式调用：

```js
function* gen() {
  console.log('hello world!')
}

let iter = gen() // 直接执行得到的是迭代器
console.log(iter) // Object [Generator] {}

iter.next() // hello world!
iter.next() // 空
```

## 二 生成器 API

### 2.1 生成器函数的声明

```js
// 生成器函数声明：*号的位置靠左靠右不影响
function* generatorFn() {}

// 生成器函数表达式
let generatorFn = function* () {}

// 作为对象字面量方法的生成器函数
let foo = {
  *generatorFn() {},
}

// 作为类实例方法的生成器函数
class Foo {
  *generatorFn() {}
}

// 作为类静态方法的生成器函数
class Bar {
  static *generatorFn() {}
}
```

注意：箭头函数不能用来定义生成器函数。

### 2.2 生成器对象

运行生成器函数会得到一个生成器对象，此时生成器对象处于暂停执行状态（suspended）。由于生成器对象实现了 Iterator 接口，因此具备 next() 方法，调用 next() 会让生成器开始、恢复执行：

```js
function* gen() {
  console.log('hello world!')
  return 'foo'
}

let g = gen() // 没有输出，只是获取得迭代器

console.log(g.next()) // 输出 helloworld，{ value: 'foo', done: true }
console.log(g.next()) // { value: undefined, done: true }
```

生成器对象的默认迭代器是自引用的：

```js
function* gen() {}

const g = gen()

console.log(g === g[Symbol.iterator]()) // true
```

### 2.2 yeild 语句

生成器最重要的地方是：可以通过 yeild 关键字让生成器开始、停止执行！

生成器函数在遇到 yield 之前会正常执行，遇到该关键字则停止执行，函数的作用域状态会被保留，重新调用 next() 之后可以恢复执行！

`yeild` 语句可以看做函数代码的分隔符，将函数拆分成了多个部分，每个 `yeild` 将会由一个 next 执行：

```js
function* gen() {
  console.log('hello world!')
  yield 'hello 1'
  yield
  yield 'hello 3'
  return 'world'
}

let y = gen()

console.log(y.next()) // hello world! { value: 'hello 1', done: false }
console.log(y.next()) // { value: undefined, done: false }
console.log(y.next()) // { value: 'hello 3', done: false }
console.log(y.next()) // { value: 'world', done: true }
console.log(y.next()) // { value: undefined, done: true }
```

yield 关键字有点像函数的中间返回语句，它生成的值会出现在 next() 方法返回的对象里，通过 yield 关键字退出的生成器函数会处在 done: false 状态；通过 return 关键字退出的生成器函数会处于 done: true 状态。

贴士：

- 生成器函数内部的执行流程会针对每个生成器对象区分作用域。在一个生成器对象上调用 next() 不会影响其他生成器的。
- yield 关键字只能在生成器函数内使用

### 2.3 参数传递

next() 方法可以传入实参，这个实参将会作为执行时对应 yield 语句的返回结果！

```js
function* gen(param) {
  console.log(param)
  console.log(yield)
  console.log(yield)
}

// 第一次调用 next() 传入的值不会被使用，因为这一次调用是为了开始执行生成器函数
// foo baz qux
let g = gen('foo')
g.next('bar')
g.next('baz')
g.next('qux')
```

yield 关键字可以同时用于输入和输出：

```js
function* gen() {
  return yield 'foo'
}
let g = gen()
console.log(g.next()) // { done: false, value: 'foo' }
console.log(g.next('bar')) // { done: true, value: 'bar' }
```

### 2.4 生成器提前终止

与迭代器类似，生成器也支持“可关闭”的概念。一个实现 Iterator 接口的对象一定有 next() 方法，还有一个可选的 return() 方法用于提前终止迭代器。生成器对象除了有这两个方法，还有第三个方法：throw()。

```js
function* generatorFn() {}
const g = generatorFn()
console.log(g) // generatorFn {<suspended>}
console.log(g.next) // f next() { [native code] }
console.log(g.return) // f return() { [native code] }
console.log(g.throw) // f throw() { [native code] }
```

return() 和 throw() 方法都可以用于强制生成器进入关闭状态：

```js
function* generatorFn() {
  for (const x of [1, 2, 3]) {
    yield x
  }
}
const g = generatorFn()
console.log(g) // generatorFn {<suspended>}
console.log(g.return(4)) // { done: true, value: 4 }
console.log(g) // generatorFn {<closed>}
```

与迭代器不同，所有生成器对象都有 return() 方法，只要通过它进入关闭状态，就无法恢复了。后续调用 next() 会显示 done: true 状态，而提供的任何返回值都不会被存储或传播：

```js
function* generatorFn() {
  for (const x of [1, 2, 3]) {
    yield x
  }
}

const g = generatorFn()
console.log(g.next()) // { done: false, value: 1 }
console.log(g.return(4)) // { done: true, value: 4 }
console.log(g.next()) // { done: true, value: undefined }
console.log(g.next()) // { done: true, value: undefined }
console.log(g.next()) // { done: true, value: undefined }
```

for-of 循环等内置语言结构会忽略状态为 done: true 的 IteratorObject 内部返回的值。

throw() 方法会在暂停的时候将一个提供的错误注入到生成器对象中。如果错误未被处理，生成器就会关闭：

```js
function* generatorFn() {
  for (const x of [1, 2, 3]) {
    yield x
  }
}
const g = generatorFn()
console.log(g) // generatorFn {<suspended>}
try {
  g.throw('foo')
} catch (e) {
  console.log(e) // foo
}
console.log(g) // generatorFn {<closed>}
```

不过，假如生成器函数内部处理了这个错误，那么生成器就不会关闭，而且还可以恢复执行。错误处理会跳过对应的 yield，因此在这个例子中会跳过一个值。比如：

```js
function* generatorFn() {
  for (const x of [1, 2, 3]) {
    try {
      yield x
    } catch (e) {}
  }
}

const g = generatorFn()
console.log(g.next()) // { done: false, value: 1}
g.throw('foo')
console.log(g.next()) // { done: false, value: 3}
```

## 三 生成器的应用

### 3.1 生成器应用

在生成器上显式调用 next() 方法作用不大，生成器对象应该作为可迭代对象，使用起来更方便：

```js
function* nTimes(n) {
  while (n--) {
    yield
  }
}

// foo foo foo
for (let _ of nTimes(3)) {
  console.log('foo')
}
```

### 3.2 生成器函数在异步中的应用

三个异步函数的实例：

```js
function fn1() {
  setTimeout(() => {
    console.log(1111)
  }, 1000)
}

function fn2() {
  setTimeout(() => {
    console.log(2222)
  }, 1100)
}

function fn3() {
  setTimeout(() => {
    console.log(3333)
  }, 1200)
}
```

如果要让其按照 fn1、fn2、fn3 的顺序执行，那么按照回调函数的做法应该是：

```js
setTimeout(){
  setTimeout(){
    setTimeout(){

    }
  }
}
```

这显然形成了回调地狱，现在利用生成器函数可以写为：

```js
function fn1() {
  setTimeout(() => {
    console.log(1111)
    iterator.next()
  }, 1000)
}

function fn2() {
  setTimeout(() => {
    console.log(2222)
    iterator.next()
  }, 1100)
}

function fn3() {
  setTimeout(() => {
    console.log(3333)
    iterator.next()
  }, 1200)
}

// 如果我们想让 fn1、fn2、fn3 按顺序执行
function* gen() {
  yield fn1()
  yield fn2()
  yield fn3()
}

let iterator = gen()
iterator.next()
```
